File: tests/system_tests/system_test.cpp
/*----------------------------------------------------------------------------*\
 (c) Crown copyright 2022 Met Office. All rights reserved.
 The file LICENCE, distributed with this code, contains details of the terms
 under which the code may be used.
\*----------------------------------------------------------------------------*/

#include <iostream>
#include "profiler.h"
#include <omp.h>
#include <mpi.h>

int main() {

    // Initialise MPI
    MPI_Init(NULL,NULL);
    MPI_Comm comm = MPI_COMM_WORLD;

    // Get current MPI rank
    int crank;
    MPI_Comm_rank(comm, &crank);

    // Begin OpenMP region
    #pragma omp parallel default(none) shared(std::cout,prof,crank)
    {
        // Get current OMP thread
        int cthread = omp_get_thread_num();

        // Profile a simple print statement
        auto prof_print = prof.start("print statement");
        #pragma omp critical
        {
            std::cout << "MPI rank " << crank << ", OMP thread " << cthread << std::endl;
        }  
        prof.stop(prof_print);
    }

    // Finish
    prof.write();
    MPI_Finalize();

    return 0;
}

File: tests/system_tests/system_test.f90
program main
    use profiler_mod
    use omp_lib
    use mpi

    implicit none

    integer :: ierr
    integer :: comm
    integer :: crank
    integer :: cthread

    integer (kind=pik) :: prof_print

    ! Initialise MPI
    call MPI_Init(ierr)
    comm = MPI_COMM_WORLD

    ! Get current MPI rank
    call MPI_Comm_rank(comm, crank, ierr)

    ! Begin OpenMP region
    !$OMP PARALLEL DEFAULT(NONE) SHARED(crank,prof_print) PRIVATE(cthread)
        
        ! Get current OMP thread
        cthread = omp_get_thread_num()

        ! Profile a simple print statement
        call profiler_start(prof_print, "print statement")
        !$OMP CRITICAL
            print *, 'MPI rank ', crank, ', OMP thread ', cthread
        !$OMP END CRITICAL
        call profiler_stop(prof_print)
        
    !$OMP END PARALLEL

    ! Finish
    call MPI_Finalize(ierr)
    call profiler_write()

end program main
